src/compiler/crystal/tools/formatter.cr

```diff
@@ -2480,11 +2480,13 @@ module Crystal
         next_token_skip_space_or_newline
       end
 
-      if @token.keyword?(:do)
+      needs_do_block = node.location.try(&.line_number) != node.end_location.try(&.line_number)
```

Comment:
I'm not entirely sure this is the correct way to detect this but it seems to work.